[Merged by Bors] - feat: lemmas about germs in topological spaces

grunweg opened the pull request:
From sphere-eversion; I'm just upstreaming it.

Co-authored by: @PatrickMassot (original author)

---

- [x] depends on: #9674

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ADedecker commented:
Could you move the import tweaks to another PR please?

ADedecker commented on Mathlib/Order/Filter/Basic.lean:
```diff
@@ -3346,6 +3347,10 @@ theorem Set.MapsTo.tendsto {α β} {s : Set α} {t : Set β} {f : α → β} (h
   Filter.tendsto_principal_principal.2 h
 #align set.maps_to.tendsto Set.MapsTo.tendsto
 
+theorem Filter.EventuallyEq.comp_fun {f g : β → γ} (h : f =ᶠ[l'] g)
```
This already exists as [Filter.EventuallyEq.comp_tendsto](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/Germ.html#Filter.EventuallyEq.comp_tendsto) (and I think the existing name is better, but that is debatable). It should absolutely be moved around here though

ADedecker commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,190 @@
+  rfl
+
+/-- The germ of functions `X → Y` at `x ∈ X` is constant w.r.t. the neighbourhood filter `𝓝 x`. -/
+def IsConstant (P : Germ (𝓝 x) Y) : Prop :=
+  P.liftOn (fun f ↦ ∀ᶠ x' in 𝓝 x, f x' = f x) <| by
+    suffices : ∀ f g : X → Y, f =ᶠ[𝓝 x] g → (∀ᶠ x' in 𝓝 x, f x' = f x) → ∀ᶠ x' in 𝓝 x, g x' = g x
+    exact fun f g hfg ↦ propext ⟨fun h ↦ this f g hfg h, fun h ↦ this g f hfg.symm h⟩
+    rintro f g hfg hf
+    refine (hf.and hfg).mono fun x' hx' ↦ ?_
+    rw [← hx'.2, hx'.1, hfg.eq_of_nhds]
```
This could be defined for any filter, right? Of course you wouldn't have the distinguished value `x` so the proof might get slightly more complicated, but I think this is something that we want in general.

grunweg replied:
Good point. In fact, my new definition makes the proofs slightly *simpler*. :-)

ADedecker commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,190 @@
+
+end Filter.Germ
+
+theorem eq_of_germ_isConstant [PreconnectedSpace X]
+    (h : ∀ x : X, (f : Germ (𝓝 x) Y).IsConstant) (x x' : X) : f x = f x' := by
+  revert x
+  erw [← eq_univ_iff_forall]
+  apply IsClopen.eq_univ _ (⟨x', rfl⟩ : {x | f x = f x'}.Nonempty)
+  refine ⟨isOpen_iff_eventually.mpr fun x hx ↦ hx ▸ h x, ?_⟩
+  rw [isClosed_iff_frequently]
+  rintro x hx
+  rcases ((h x).and_frequently hx).exists with ⟨x'', H⟩
+  exact H.1.symm.trans H.2
+
+theorem eq_of_germ_isConstant_on {s : Set X}
+    (h : ∀ x ∈ s, (f : Germ (𝓝 x) Y).IsConstant) (hs : IsPreconnected s) {x' : X} (x_in : x ∈ s)
+    (x'_in : x' ∈ s) : f x = f x' := by
+  haveI := isPreconnected_iff_preconnectedSpace.mp hs
+  let F : s → Y := f ∘ (↑)
+  change F ⟨x, x_in⟩ = F ⟨x', x'_in⟩
+  apply eq_of_germ_isConstant
+  rintro ⟨x, hx⟩
+  have : ContinuousAt ((↑) : s → X) ⟨x, hx⟩ := continuousAt_subtype_val
+  exact this (h x hx)
```
These should be proved using [IsLocallyConstant.apply_eq_of_isPreconnected](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/LocallyConstant/Basic.html#IsLocallyConstant.apply_eq_of_isPreconnected), unless this is a too heavy import (which would be a shame)

grunweg replied:
I just tried this: for the first lemma, that's easy. I wouldn't mind a mind with the second one.

grunweg replied:
The second one is taking me further afield that I'd like: I've left a proof_wanted with some hints.

ADedecker commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,190 @@
+      rwa [Germ.coe_eq.mpr (EventuallyEq.symm hy')]
+    fun f f' hff' ↦ propext <| forall_congr' fun _ ↦ ⟨this f f' hff', this f' f hff'.symm⟩
+
+theorem Filter.Eventually.germ_congr
+    {P : Germ (𝓝 x) Y → Prop} (hf : P f) (h : ∀ᶠ z in 𝓝 x, g z = f z) : P g := by
+  convert hf using 1
+  exact Germ.coe_eq.mpr h
```
This looks a bit weird to me, as it seems like a very specialized combination of already existing facts. Is it useful for dot notation? Could you link some uses in the sphere eversion project so I see where it's useful?

grunweg replied:
All uses of `germ_congr{,_set}` are in [InductiveConstruction](https://github.com/leanprover-community/sphere-eversion/blob/master/SphereEversion/InductiveConstructions.lean). There are four places overall; it might be easiest to just use the search function.

grunweg replied:
I looked now: the one use of germ_congr is easy to inline. I would suggest I remove `germ_congr` and keep `germ_congr_set`. What do you think?

grunweg replied:
I inlined `germ_congr` into `germ_congr_set`.

ADedecker commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,190 @@
+`forall_restrict_germ_predicate_iff` for this equivalence. -/
+def RestrictGermPredicate (P : ∀ x : X, Germ (𝓝 x) Y → Prop)
+    (A : Set X) : ∀ x : X, Germ (𝓝 x) Y → Prop := fun x φ ↦
+  Quotient.liftOn' φ (fun f ↦ x ∈ A → ∀ᶠ y in 𝓝 x, P y f)
```
This should use [Filter.Germ.liftOn](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/Germ.html#Filter.Germ.liftOn) instead of `Quotient.liftOn'`

ADedecker commented on Mathlib/Order/Filter/Germ.lean:
```diff
@@ -257,6 +257,11 @@ theorem compTendsto'_coe (f : Germ l β) {lc : Filter γ} {g : γ → α} (hg :
+
+theorem Filter.Tendsto.congr_germ {f g : β → γ} {l : Filter α} {l' : Filter β} (h : f =ᶠ[l'] g)
+    {φ : α → β} (hφ : Tendsto φ l l') : (f ∘ φ : Germ l γ) = g ∘ φ :=
+  @Quotient.sound _ (l.germSetoid γ) _ _ (hφ h)
```
The proof should use [Filter.EventuallyEq.germ_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/Germ.html#Filter.EventuallyEq.germ_eq) instead of `Quotient.sound`.

I'm also wondering wether this deserves a lemma, is `(h.comp_tendsto hφ).germ_eq` not short enough?

grunweg replied:
Good question! Let me look at sphere-eversion...

grunweg replied:
It is unused there. In other words: I don't have a strong opinion and am happy to delete this.

grunweg commented:
Some loose ends remaining, but I think this is ready for another look.

grunweg commented:
I have addressed all comments.

sgouezel commented on Mathlib/Order/Filter/Germ.lean:
```diff
@@ -257,6 +275,17 @@ theorem compTendsto'_coe (f : Germ l β) {lc : Filter γ} {g : γ → α} (hg :
+    {φ : α → β} (hφ : Tendsto φ l l') : (f ∘ φ : Germ l γ) = g ∘ φ :=
+  EventuallyEq.germ_eq (h.comp_tendsto hφ)
+
+/-- If a germ `f : Germ l β` is constant, where `l : Filter α`,
+and a function `g : γ → α` tends to `l` along `lc : Filter γ`,
+the germ of the composition `f ∘ g` is also constant. -/
+proof_wanted isConstant_compTendsto {f : Germ l β} {lc : Filter γ} {g : γ → α}
+    (_hf : (f : Germ l β).IsConstant) (_hg : Tendsto g lc l) : (f.compTendsto g _hg).IsConstant
```
```suggestion
lemma isConstant_comp_tendsto {lc : Filter γ} {g : γ → α}
    (hf : (f : Germ l β).IsConstant) (hg : Tendsto g lc l) : IsConstant (f ∘ g : Germ lc β ) := by
  rcases hf with ⟨b, hb⟩
  exact ⟨b, hb.comp_tendsto hg⟩

/-- If a germ `f : Germ l β` is constant, where `l : Filter α`,
and a function `g : γ → α` tends to `l` along `lc : Filter γ`,
the germ of the composition `f ∘ g` is also constant. -/
lemma isConstant_compTendsto {f : Germ l β} {lc : Filter γ} {g : γ → α}
    (hf : f.IsConstant) (hg : Tendsto g lc l) : (f.compTendsto g hg).IsConstant := by
  rcases Quotient.exists_rep f with ⟨f, rfl⟩
  exact isConstant_comp_tendsto hf hg
```

sgouezel commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,192 @@
+  exact ha
+
+-- should follow from `isConstant_compTendsto` specialised to the neigbourhood filter
+proof_wanted isConstant_comp_subtype {s : Set X} {f : X → Y} {x : s}
```
This one is now `isConstant_comp_tendsto hf continuousAt_subtype_val`. Maybe move it a few lines up, in the namespace `Filter.Germ`?

sgouezel commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,192 @@
+    ((f ∘ Subtype.val : s → Y) : Germ (𝓝 x) Y).IsConstant
+
+-- move to `LocallyConstant/Basic.lean` once proven
+proof_wanted IsLocallyConstant.of_comp_of_inducing
```
This one looks wrong. If `i` is the inclusion of a point in a space, say, then `f o i` is locally constant whatever the function `f`, but there is no reason why `f` itself should be locally constant.

grunweg replied:
Good point. I clearly went to far in extracting intermediate lemmas from `eq_of_germ_isConstant_on` (which *did* have a proof in sphere-eversion). Deleted.

sgouezel commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,192 @@
+proof_wanted IsLocallyConstant.of_germ_isConstantOn_of_preconnected {s : Set X} [TopologicalSpace Y]
+    (_hs : IsPreconnected s) (_h : ∀ x ∈ s, (f : Germ (𝓝 x) Y).IsConstant) : IsLocallyConstant f
+
+theorem eq_of_germ_isConstant [i: PreconnectedSpace X]
```
```suggestion
theorem eq_of_germ_isConstant [i : PreconnectedSpace X]
```

sgouezel commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,192 @@
+  {i : Z → X} (_hi : Inducing i) (_hF : IsLocallyConstant (f ∘ i)) : IsLocallyConstant f
+
+-- use previous lemma: `Subtype.val` is inducing
+proof_wanted IsLocallyConstant.of_germ_isConstantOn_of_preconnected {s : Set X} [TopologicalSpace Y]
```
This one is also wrong: you only have assumptions on `s`, you can not have a global conclusion.

grunweg commented:
Thank you for the careful look; I have addressed all comments.
- I'm not convinced whether `eq_of_germ_isConstant_on` is too useful (happy to remove it if you prefer)
- I'm happy to hear suggestions where `Subring.orderedSubtype` could be placed.

sgouezel commented on Mathlib/Order/Filter/Germ.lean:
```diff
@@ -257,6 +272,24 @@ theorem compTendsto'_coe (f : Germ l β) {lc : Filter γ} {g : γ → α} (hg :
+  EventuallyEq.germ_eq (h.comp_tendsto hφ)
+
+lemma isConstant_comp_tendsto {lc : Filter γ} {g : γ → α}
+    (hf : (f : Germ l β).IsConstant) (hg : Tendsto g lc l) : IsConstant (f ∘ g : Germ lc β ) := by
```
```suggestion
    (hf : (f : Germ l β).IsConstant) (hg : Tendsto g lc l) : IsConstant (f ∘ g : Germ lc β) := by
```

sgouezel commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,182 @@
+end Filter.Germ
+
+/-- If the germ of `f` w.r.t. each `𝓝 x` is constant, `f` is locally constant. -/
+private lemma IsLocallyConstant.of_germ_isConstant (h : ∀ x : X, (f : Germ (𝓝 x) Y).IsConstant) :
```
This lemma shouldn't be private, it could be useful more generally.

sgouezel commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,182 @@
+  (IsLocallyConstant.of_germ_isConstant h).apply_eq_of_isPreconnected
+    (preconnectedSpace_iff_univ.mp i) (by trivial) (by trivial)
+
+proof_wanted eq_of_germ_isConstant_on {s : Set X} (_h : ∀ x ∈ s, (f : Germ (𝓝 x) Y).IsConstant)
+    (_hs : IsPreconnected s) {x' : X} (_x_in : x ∈ s) (_x'_in : x' ∈ s) : f x = f x'
```
```suggestion
lemma eq_of_germ_isConstant_on {s : Set X} (h : ∀ x ∈ s, (f : Germ (𝓝 x) Y).IsConstant)
    (hs : IsPreconnected s) {x' : X} (x_in : x ∈ s) (x'_in : x' ∈ s) : f x = f x' := by
  let i : s → X := fun x ↦ x
  show (f ∘ i) (⟨x, x_in⟩ : s) = (f ∘ i) (⟨x', x'_in⟩ : s)
  have : PreconnectedSpace s := Subtype.preconnectedSpace hs
  exact eq_of_germ_isConstant (fun y ↦ Germ.isConstant_comp_subtype (h y y.2)) _ _
```

sgouezel commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,182 @@
+
+end Filter.Germ
+
+/-- The inclusion `S → R` of a subring, as an ordered ring homomorphism. -/
```
I would just create a new file `RingTheory.Subring.Ordered`.

sgouezel commented on Mathlib/Topology/Germ.lean:
```diff
@@ -0,0 +1,182 @@
+
+/-- The inclusion `S → R` of a subring, as an ordered ring homomorphism. -/
+-- xxx: OrderedRing has no morphisms, OrderedRingHom no subtypes -> which file is a good place?
+def _root_.Subring.orderedSubtype {R} [OrderedRing R] (s : Subring R) : s →+*o R :=
+  { s.subtype with monotone' := fun _ _ h ↦ h }
```
```suggestion
def Subring.orderedSubtype {R} [OrderedRing R] (s : Subring R) : s →+*o R where
  __ := s.subtype 
  monotone' := fun _ _ h ↦ h
```
is now the recommended spelling.

grunweg commented:
Thanks again for the comments, which I have addressed.
awaiting-review

sgouezel commented:
Can you fix the build by following the linter instructions?

grunweg commented:
Fixed the imports linter and merged master.
There's a freshly-generated `Subrings/Order.lean`; I could move the new material to that file instead. (Note that `Ordered.lean` has slightly heavier imports, though.)
awaiting-review

grunweg commented:
Update: Now it builds again, `lake exe shake` was a bit overzealous.
I'm not sure if my tweak to `noshake.json` worked (locally, it didn't --- reading [this](https://github.com/leanprover-community/mathlib4/pull/9830#issuecomment-1897969615), it seemed like it should). If things are amiss, help is welcome.

sgouezel commented:
Yes, please move it to `Subring.Order`, it looks like a very good fit even if it means adding some imports there.

grunweg commented:
Thanks; updated. (CI has already passed the build and lint steps.)

sgouezel commented:
bors r+
Thanks!

sgouezel commented:
bors r-

sgouezel commented:
bors r+
